napi-inl.h

```diff
@@ -3697,10 +3697,26 @@ inline napi_value ObjectWrap<T>::ConstructorCallbackWrapper(
   }
 
   T* instance;
-  napi_value wrapper = details::WrapCallback([&] {
+  napi_value wrapper = details::WrapCallback([&] () -> napi_value {
     CallbackInfo callbackInfo(env, info);
+    callbackInfo.zombie = new Zombie();
+#ifdef NAPI_CPP_EXCEPTIONS
+    try {
+      instance = new T(callbackInfo);
+      return callbackInfo.This();
+    }
+    catch (...) {
+      callbackInfo.zombie->isZombie = true;
+      throw;
+    }
+#else
     instance = new T(callbackInfo);
+    if (callbackInfo.Env().IsExceptionPending()) {
+      callbackInfo.zombie->isZombie = true;
+      return nullptr;
+    }
     return callbackInfo.This();
+#endif
   });
```

Comment:
I think it might be best to free `zombie` here and not pass it as the `hint` to the finalize callback, because, if `napi_wrap()` fails, that callback will not be called.